Fix catalog entry sort

fdiaz opened the pull request:
Closes https://github.com/SwiftGen/SwiftGenKit/issues/56

AliSoftware commented:
Note that this is a duplicate of #55 so whichever of those two will be validated and finished first will invalidate the other one.

_PS: If you wish, it could be a nice courtesy to also credit the original author of #55 in your CHANGELOG entry too if you used #55 as inspiration and this #56 is just a continuation of their work to finish the work_ 😉

fdiaz commented:
You're totally right @AliSoftware ! Updated the CHANGELOG. I didn't wanna build on top of #55 since this is a slightly different approach.

#55 modifies `AssetsCatalogContext.swift` and this PR modifies `AssetsCatalogParser.swift`.

I took this road since the underlying problem is with the `children()` function. So this sorts the ouput of anyone using `process(folder: Path, withPrefix prefix: String = "")`

AliSoftware commented on Tests/SwiftGenKitTests/CatalogEntryTests.swift:
```diff
@@ -0,0 +1,33 @@
+final class CatalogEntryTests: XCTestCase {
+  func testGroupEntry_IsComparableAlphabetically() {
+    let entryA = Catalog.Entry.group(name: "A", items: [])
+    let entryB = Catalog.Entry.group(name: "B", items: [])
```
Could be nice to also add some tests using different letter case to check if that comparability is case-sensitive or not (but whether it is or not, that it always sorts `["A","a"].map { Catalog.Entry.group(name: $0, items: []) }` in the same order

fdiaz replied:
Now that I think about it, I should also test that two groups with the same name order its items correctly.

I'll work on that 👍

AliSoftware commented on Tests/SwiftGenKitTests/CatalogEntryTests.swift:
```diff
@@ -0,0 +1,33 @@
+    let entryA = Catalog.Entry.group(name: "A", items: [])
+    let entryB = Catalog.Entry.group(name: "B", items: [])
+
+    XCTAssertEqual(entryA < entryB, true)
```
The expected usage of `XCTAssertEqual` is to take 2 parameters to equate, like `XCTAssertEqual(entryA, entryB)`.

If you want to check if an expression, like `entryA < entryB`, is true, prefer `XCTAssert` or `XCTAssertTrue`.
_(Some even prefer using `XCTAssertLessThan(entryA, entryB)` but this last one being less commonly known, `XCTAssert` or `XCTAssertTrue` are generally more fitting)_

AliSoftware commented on Sources/Parsers/AssetsCatalogParser.swift:
```diff
@@ -34,18 +34,48 @@ extension Catalog.Entry: Equatable {
+      return .orderedDescending
+    }
+
+    return compare(Array(lhs.dropFirst()), Array(rhs.dropFirst()))
```
I'm not a fan of using this recursive approach as it can make the stack grow quite quickly for a group containing 100s of images…

What about:
```swift
private static func compare(_ lhs: [Catalog.Entry], _ rhs: [Catalog.Entry]) -> ComparisonResult {
  for pair in zip(lhs, rhs) {
    if pair.0 < pair.1 { return .orderedAscending }
    if pair.0 > pair.1 { return .orderedDescending }
  }
  // If we arrive a this point, this means all pairs have been considered equal
  // but one of the list might be longer than the other
  if lhs.count < rhs.count { return .orderedAscending }
  if lhs.count > rhs.count { return .orderedDescending }
  return .orderedSame
}
```

fdiaz replied:
That's a good point and a much cleaner implementation. Updating 👍

AliSoftware replied:
I even wonder if the last part (last 3 lines) couldn't be reduced to `return compare(lhs.count, rhs.count)`?
(Don't we have a `compare(Int,Int) -> ComparisonResult` in the stdlib?)

AliSoftware commented on Sources/Parsers/AssetsCatalogParser.swift:
```diff
@@ -82,7 +133,7 @@ extension AssetsCatalogParser {
   func process(folder: Path, withPrefix prefix: String = "") -> [Catalog.Entry] {
     return (try? folder.children().flatMap {
```
@fdiaz I just realised: wouldn't it be simpler instead of all the `Equatable` + `Comparable` implementations you added… to just sort the output of `children()` directly? i.e. apply `folder.children().sorted(by: <).flatMap { … }` to children, instead of using `folder.children().flatMap { … }.sorted(by: <)` and sort the the final, recursively-constructed structure?

Which would mean in practice removing all the code that you added above and a simpler and mode performant solution (because you don't have to traverse the recursive structure twice) in the end, right?

fdiaz replied:
Oh wow, you're right. I didn't notice that `Path` was already `Comparable`. I'll just go back with all these changes and add that little change instead. Nice catch!

fdiaz committed:
Add CHANGELOG entry

fdiaz committed:
Sort children output when parsing folder

The children() function uses under the hood
contentsOfDirectory(atPath:) https://developer.apple.com/documentation/foundation/filemanager/1414584-contentsofdirectory

This array returned by that function has an undefined order.

djbe commented on CHANGELOG.md:
```diff
@@ -6,7 +6,10 @@
 ### Bug Fixes
 
-_None_
+* Fixes an issue in High Sierra where the output of the processed Catalog Entries was not ordered alphabetically.
```
This seems to be missing the 2 spaces at the end?

djbe reviewed (changes requested):
We should probably also change our tests to run on the latest Xcode, and the latest macOS (if possible), to ensure that this is actually fixed (and remains so).

AliSoftware commented:
Good point @djbe our scripts would probably need updating, or more precisely the `.circle.yml` configuration should target 10.13 at some point.

@fdiaz any news on the rebase we spoke about on slack simplifying this whole PR?

djbe commented:
What simplification? This PR boils down to 1 line of code 😄

AliSoftware commented:
Oh I missed the notification telling me that @fdiaz did in fact already commit the simplification we talked about about sorting `children()` directly (idk if you saw the PR before that @djbe but it was sorting the output of `process()` instead, having to do a recursive sort and having to implement `Equatable`+`Comparable` on `Catalog.Entry` and all.

Seems like the Circle tests are still failing though.

djbe commented:
Yeah, I partially followed it, will fix the contexts.

djbe committed:
align submodule

djbe committed:
spaces

nafu commented:
Cool 💚

fdiaz commented:
Thanks for finishing this up @djbe 👍 

I'm not sure why the tests were still failing though? Any context on that? I was about to take a look on High Sierra to see what was going on with that.

djbe commented:
There is a slight difference between the (built in) sort-order in Sierra, and the sorting we now apply. Nothing much to worry about.
